File: web/src/lib/api/library.ts
import { request } from './client';
import type {
  CreateLibraryItemRequest,
  LibraryContentResponse,
  LibraryItem,
  ListLibraryItemsParams,
  ListLibraryItemsResponse,
  UpdateLibraryItemRequest,
} from './types';

export function listLibraryItems(
  params: ListLibraryItemsParams = {},
  signal?: AbortSignal,
): Promise<ListLibraryItemsResponse> {
  const query = new URLSearchParams();

  if (params.status) query.set('status', params.status);
  if (params.q?.trim()) query.set('q', params.q.trim());
  if (params.tag?.trim()) query.set('tag', params.tag.trim());
  if (params.read !== undefined) query.set('read', String(params.read));
  if (params.starred !== undefined) query.set('starred', String(params.starred));
  if (params.page !== undefined) query.set('page', String(params.page));
  if (params.pageSize !== undefined) query.set('page_size', String(params.pageSize));

  const suffix = query.size ? `?${query.toString()}` : '';
  return request<ListLibraryItemsResponse>(`/library${suffix}`, { method: 'GET', signal });
}

export function createLibraryItem(payload: CreateLibraryItemRequest): Promise<LibraryItem> {
  return request<LibraryItem>('/library', { body: payload, method: 'POST' });
}

export function getLibraryItem(uid: string, signal?: AbortSignal): Promise<LibraryItem> {
  return request<LibraryItem>(`/library/${encodeURIComponent(uid)}`, { method: 'GET', signal });
}

export function getLibraryContent(
  uid: string,
  signal?: AbortSignal,
): Promise<LibraryContentResponse> {
  return request<LibraryContentResponse>(`/library/${encodeURIComponent(uid)}/content`, {
    method: 'GET',
    signal,
  });
}

export function retryLibraryItem(uid: string, signal?: AbortSignal): Promise<LibraryItem> {
  return request<LibraryItem>(`/library/${encodeURIComponent(uid)}/retry`, {
    method: 'POST',
    signal,
  });
}

export function updateLibraryItem(
  uid: string,
  payload: UpdateLibraryItemRequest,
): Promise<LibraryItem> {
  return request<LibraryItem>(`/library/${encodeURIComponent(uid)}`, {
    body: payload,
    method: 'PATCH',
  });
}

export function deleteLibraryItem(uid: string): Promise<void> {
  return request<void>(`/library/${encodeURIComponent(uid)}`, { method: 'DELETE' });
}

File: src/api/library.rs
use axum::{
    Json, Router,
    extract::{Path, State},
    http::StatusCode,
    routing::{get, post},
};
use serde::Deserialize;

use crate::{
    api::{ApiJson, ApiQuery},
    error::AppResult,
    library::{
        self, CreateLibraryItemRequest, LibraryContent, LibraryItem, LibraryStatus,
        ListLibraryItemsOptions, ListLibraryItemsResponse, UpdateLibraryItemRequest,
    },
    state::AppState,
    workspace::RequestContext,
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/library", get(list).post(create))
        .route("/library/{uid}", get(get_one).patch(update).delete(delete))
        .route("/library/{uid}/content", get(get_content))
        .route("/library/{uid}/retry", post(retry_fetch))
}

#[derive(Deserialize)]
#[serde(default, deny_unknown_fields)]
struct ListLibraryItemsQuery {
    status: LibraryStatus,
    q: Option<String>,
    tag: Option<String>,
    read: Option<bool>,
    starred: Option<bool>,
    page: u32,
    page_size: u32,
}

impl Default for ListLibraryItemsQuery {
    fn default() -> Self {
        Self {
            status: LibraryStatus::Active,
            q: None,
            tag: None,
            read: None,
            starred: None,
            page: 1,
            page_size: 30,
        }
    }
}

async fn list(
    State(state): State<AppState>,
    context: RequestContext,
    ApiQuery(query): ApiQuery<ListLibraryItemsQuery>,
) -> AppResult<Json<ListLibraryItemsResponse>> {
    Ok(Json(
        library::list(
            state.pool(),
            context.workspace_id,
            ListLibraryItemsOptions {
                status: query.status,
                query: query.q.as_deref(),
                tag: query.tag.as_deref(),
                read: query.read,
                starred: query.starred,
                page: query.page,
                page_size: query.page_size,
            },
        )
        .await?,
    ))
}

async fn create(
    State(state): State<AppState>,
    context: RequestContext,
    ApiJson(request): ApiJson<CreateLibraryItemRequest>,
) -> AppResult<(StatusCode, Json<LibraryItem>)> {
    let outcome = library::create(
        state.pool(),
        context.workspace_id,
        context.user_id,
        request,
        state.clock().now().timestamp_millis(),
    )
    .await?;
    let status = if outcome.created {
        StatusCode::CREATED
    } else {
        StatusCode::OK
    };
    Ok((status, Json(outcome.item)))
}

async fn get_one(
    State(state): State<AppState>,
    context: RequestContext,
    Path(uid): Path<String>,
) -> AppResult<Json<LibraryItem>> {
    Ok(Json(
        library::get(state.pool(), context.workspace_id, &uid).await?,
    ))
}

async fn get_content(
    State(state): State<AppState>,
    context: RequestContext,
    Path(uid): Path<String>,
) -> AppResult<Json<LibraryContent>> {
    Ok(Json(
        library::get_content(state.pool(), context.workspace_id, &uid).await?,
    ))
}

async fn retry_fetch(
    State(state): State<AppState>,
    context: RequestContext,
    Path(uid): Path<String>,
) -> AppResult<(StatusCode, Json<LibraryItem>)> {
    Ok((
        StatusCode::ACCEPTED,
        Json(
            library::retry_fetch(
                state.pool(),
                context.workspace_id,
                &uid,
                state.clock().now().timestamp_millis(),
            )
            .await?,
        ),
    ))
}

async fn update(
    State(state): State<AppState>,
    context: RequestContext,
    Path(uid): Path<String>,
    ApiJson(request): ApiJson<UpdateLibraryItemRequest>,
) -> AppResult<Json<LibraryItem>> {
    Ok(Json(
        library::update(
            state.pool(),
            context.workspace_id,
            &uid,
            request,
            state.clock().now().timestamp_millis(),
        )
        .await?,
    ))
}

async fn delete(
    State(state): State<AppState>,
    context: RequestContext,
    Path(uid): Path<String>,
) -> AppResult<StatusCode> {
    library::delete(state.pool(), context.workspace_id, &uid).await?;
    Ok(StatusCode::NO_CONTENT)
}
